src/adapters/MorphoMarketV1Adapter.sol

```diff
@@ -70,22 +79,61 @@ contract MorphoMarketV1Adapter is IMorphoMarketV1Adapter {
+        require(msg.sender == IVaultV2(parentVault).curator(), NotAuthorized());
+        require(burnSharesExecutableAt[marketId] == 0, AlreadyPending());
+        burnSharesExecutableAt[marketId] =
+            block.timestamp + IVaultV2(parentVault).timelock(IVaultV2.removeAdapter.selector);
+        emit SubmitBurnShares(marketId, burnSharesExecutableAt[marketId]);
+    }
+
+    function revokeBurnShares(bytes32 marketId) external {
+        require(
+            msg.sender == IVaultV2(parentVault).curator() || IVaultV2(parentVault).isSentinel(msg.sender),
+            NotAuthorized()
+        );
+        require(burnSharesExecutableAt[marketId] != 0, NotPending());
+        burnSharesExecutableAt[marketId] = 0;
+        emit RevokeBurnShares(marketId);
+    }
+
+    /// @dev Deallocate 0 from the vault after burning shares to update the allocation there.
+    function burnShares(bytes32 marketId) external {
+        require(burnSharesExecutableAt[marketId] != 0, NotTimelocked());
+        require(block.timestamp >= burnSharesExecutableAt[marketId], TimelockNotExpired());
+        burnSharesExecutableAt[marketId] = 0;
+        uint256 supplySharesBefore = positions[marketId].supplyShares;
+        positions[marketId].supplyShares = 0;
+        emit BurnShares(marketId, supplySharesBefore);
+    }
+
     /// @dev Does not log anything because the ids (logged in the parent vault) are enough.
     /// @dev Returns the ids of the allocation and the change in allocation.
     function allocate(bytes memory data, uint256 assets, bytes4, address) external returns (bytes32[] memory, int256) {
         MarketParams memory marketParams = abi.decode(data, (MarketParams));
         require(msg.sender == parentVault, NotAuthorized());
         require(marketParams.loanToken == asset, LoanAssetMismatch());
+        require(marketParams.irm == adaptiveCurveIrm, IrmMismatch());
+        bytes32 marketId = Id.unwrap(marketParams.id());
+        MarketPosition storage position = positions[marketId];
+
+        uint256 mintedShares;
+        if (assets > 0) {
+            (, mintedShares) = IMorpho(morpho).supply(marketParams, assets, 0, address(this), hex"");
+            require(mintedShares >= assets, SharePriceAboveOne());
+            position.supplyShares += uint128(mintedShares);
+        }
 
-        if (assets > 0) IMorpho(morpho).supply(marketParams, assets, 0, address(this), hex"");
+        uint256 oldAllocation = position.allocation;
+        uint256 _newAllocation = newAllocation(marketId);
+        updateList(marketId, oldAllocation, _newAllocation);
+        position.allocation = uint128(_newAllocation);
 
-        uint256 oldAllocation = allocation(marketParams);
-        uint256 newAllocation = MorphoBalancesLib.expectedSupplyAssets(IMorpho(morpho), marketParams, address(this));
-        updateList(marketParams, oldAllocation, newAllocation);
+        emit Allocate(marketId, _newAllocation, mintedShares);
```

Comment:
We log the new value for the allocation & the delta for the shares, does it make sense? (same for deallocate)